crates/core/src/metadata_store/providers/objstore/version_repository.rs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use bytes::Bytes;
+
+use restate_types::errors::GenericError;
+use restate_types::Version;
+
+#[derive(Debug, thiserror::Error)]
+pub enum VersionRepositoryError {
+    #[error("Version {0} already exists")]
+    AlreadyExists(Version),
+    #[error("Version {0} not found")]
+    VersionNotFound(Version),
+    #[error("Network {0}")]
+    Network(GenericError),
+}
+
+#[async_trait::async_trait]
+pub(crate) trait VersionRepository: Sync + Send + 'static {
+    async fn get_latest_version(&self) -> Result<Option<Version>, VersionRepositoryError>;
+
+    async fn try_create_version(
+        &self,
+        version: Version,
+        leaf: Bytes,
+    ) -> Result<(), VersionRepositoryError>;
+
+    async fn get_version(&self, version: Version) -> Result<Bytes, VersionRepositoryError>;
```

Comment:
nit: `get_latest_version` returns a version and `get_version` returns the payload while both sound quite similar.